src/ts_catalog/continuous_agg.h

```diff
@@ -65,8 +65,15 @@ typedef struct ContinuousAggsBucketFunction
 	char *name;
 	/* `bucket_width` argument of the function */
 	Interval *bucket_width;
-	/* `origin` argument of the function provided by the user */
-	char *origin;
+	/*
+	 * Custom origin value. If not specified, stores infinity.
+	 *
+	 * Actually, it can store Timestamp, TimestampTz or Date. Timestamp and
+	 * TimestampTz are the same in PostgreSQL so we can get away with this.
+	 * Date representation is different but it can be converted to Timestamp
+	 * with date_timestamp() without any loss of information.
+	 */
+	Timestamp origin;
```

Comment:
It does not make sense to store `origin` as a `Timestamp` type when it can actually be any of three types and a `Datum` is more appropriate. PostgreSQL is using type-punning quite heavily but it offers the guarantee that you can convert a type (e.g., `Timestamp`) to a `Datum` and back, but it offers no guarantees that you can convert between types.  Use that guarantee instead to ensure that the conversion will always work, even if PostgreSQL makes changes. It will even properly handle conversions from PostgreSQL types to references and back, if that turns out to be the case for some platform.

```suggestion
	Datum origin;
```

Reply:
That's actually a very good point, thanks. I'll try to refactor the code accordingly 👍

Reply:
On second thought, I just confused everyone with a wrong comment. What origin **actually** stores is always Timestamp in UTC, regardless of whether a CAGG works with Dates, Timestamp, or TimestampTZs. The actual problem is that there are places where I treated it as TimestampTz. I will update the comment and refactor the code according to @mfundul and @fabriziomello notes.

Reply:
I am still confused about this. I've been looking at this definition:
```
CREATE FUNCTION timescaledb_experimental.time_bucket_ng(bucket_width INTERVAL, ts TIMESTAMPTZ, origin TIMESTAMPTZ, timezone TEXT) RETURNS TIMESTAMPTZ
    AS '@MODULE_PATHNAME@', 'ts_time_bucket_ng_timezone_origin' LANGUAGE C IMMUTABLE PARALLEL SAFE STRICT;
```
In this case don't you have to push down to the data nodes the `TIMESTAMPTZ` origin instead of a `TIMESTAMP` origin? I imagine the data node will execute `ts_time_bucket_ng_timezone_origin()` with a `TIMESTAMPTZ` origin argument.

Reply:
Internally `origin` is always stored as Timestamp in UTC (Note that the previous comment I added to the code was wrong, check out the new comment). Dates and TimestampTz can both be converted to Timestamp in UTC without any data loss. The timezone is stored separately. So between data nodes `origin` can be serialized/deserialized using simply timestamp_in / timestamp_out.  See bucket_function_serialize() https://github.com/timescale/timescaledb/pull/4045/files#diff-559a60cc746cb2fe273d872b89dc6f398ae14c3a9775f38a46406fedac0aa930R425